### ISSUE TITLE
Make `+` and `*` variadic

### ISSUE THREAD

**Issue by ggevay** [MaterializeInc/materialize#13931]
#### What does the transformation do?
Currently, if the user writes `1+2+3+4` we represent it as `((1+2)+3)+4`, because `+` can have exactly 2 args. We could make `+` and `*` `VariadicFunc`.

(This is similar to https://github.com/MaterializeInc/materialize/pull/13731, but it should be much easier to do.)

#### Why should we add it?
- Avoid stack overflow in the optimizer if the user writes a long chain of additions or multiplications, e.g., `1+2+3+...+10000`.
- Slightly faster evaluation for long chains of additions or multiplications.

#### When would it be good to have?
When there are long chains of additions or multiplications in a query.

#### When would it be ineffectual?
No long chains of additions or multiplications in a query.

#### When would be bad to have?
Never.

#### In the worst case, how would it degrade performance?
It doesn't.

#### What is the benefit of the transformation?
Slightly faster evaluation for long chains of additions or multiplications.

#### What is the overhead?
None.

#### When would the transformation be worthwhile?
Always.

#### How often will users input a query that can be helped by the transformation?
Rarely. Maybe some generated code?

#### Among the queries that can be helped, are there subclasses of queries that users will input more often?
I don't know.

### Proposed implementation

Be careful that AST-level changes might also be needed, to avoid having a stack overflow already before introducing a call to the `VariadicFunc` around `plan_op`. (https://github.com/MaterializeInc/materialize/pull/13731/ didn't pay attention to this, because there the main point was the large `OR` chains introduced by `IN (1,2,3,...,10000)`.)

**Comment by aalexandrov** [comment]
As part of wrapping this up, we should also document somewhere prominently on the `BinaryFunc` enum that associative binary functions should not be put there so we don't have to go over similar effort in the future.

**Comment by frankmcsherry** [comment]
I may be totally crazy, but I don't think `+` and `*` are associative or commutative. In "math" they are, but in SQL they are not because of overflow errors.

**Comment by aalexandrov** [comment]
That's true, maybe we should re-purpose this issue to the `BitOr~` and `BitAnd` variants.

Regarding `Add~` and `Mul~`, we could still reduce the expected IR size if we represent left left-deep trees of function applications as a single variadic funciton application. If the eval method for the new variadic functions aggregates left to right and bails immediately on overflow, the operational semantics will be consistent with the current state.
